File: utilities/maths/upper_index.m
function m_ans = upper_index (arr, val)
% Return the largest index m such that arr(m)<=val given a strictly monotonically increasing array arr
%
%   >> m = upper_index (arr,val)
%
% If val<arr(1) then m=0
% val can be a scalar or an array

% Actually, the use of histc appears to give the correct result even if not *strictly* monotonic

n=numel(arr);
m=numel(val);

% Return if array has zero length:
if n==0
    m_ans=zeros(size(val));
    return
end

% Use matlab intrinsic function to get result if large enough array of test values
if m<16 || n==1 || 31*m*log(n)<n+2500
    m_ans=zeros(size(val));
    for i=1:m
        % find extremal cases, otherwise binary chop to find solution
        if (arr(n) <= val(i))
            m_ans(i) = n;
        elseif (arr(1) > val(i))
            m_ans(i) = 0;
        else
            ml = 1;
            mh = n;
            mm = floor((ml+mh+0.001)/2);
            while mm ~= ml
                if (arr(mm) > val(i))
                    mh = mm;
                else
                    ml = mm;
                end
                mm = floor((ml+mh+0.001)/2);
            end
            m_ans(i) = ml;
        end
    end
    
else
    [dummy,m_ans]=histc(val,arr(:)');
    m_ans(val>arr(end))=n;
    
end

File: external_code/Fortran/source/maths/upper_index.f90
	function upper_index_dp (arr, val)
	use type_definitions
	implicit none
!-----------------------------------------------------------------------------------------------------------------------------------
!
!	Given monotonically increasing array ARR the function returns the largest index M
!     ARR(M) =< VAL
!
!	If no such M (i.e. ARR(1) > VAL) then M=0
!
!
!-----------------------------------------------------------------------------------------------------------------------------------
!	T.G. Perring		2002-08-15		First formal release
!
!
!-----------------------------------------------------------------------------------------------------------------------------------
	real(DP), intent(in) :: arr(:), val
	integer(I4B) :: upper_index_dp

	integer(I4B) :: n, ml, mm, mh

	n = size(arr)

! return if array has zero length:
	if (n == 0) then
		upper_index_dp = 0
		return
	endif

! find extremal cases:
	if (arr(n) <= val) then
		upper_index_dp = n
		return
	else if (arr(1) > val) then
		upper_index_dp = 0
		return
	endif

! binary chop to find solution
	ml = 1
	mh = n
 10 mm = (ml+mh)/2
	if (mm == ml) then
		upper_index_dp = ml
		return
	endif
	if (arr(mm) > val) then
		mh = mm
	else
		ml = mm
	endif
	goto 10

	end
!---------------------------------------------------------------------------------------------
	function upper_index_sp (arr, val)
	use type_definitions
	implicit none
!
!	Given monotonically increasing array ARR the function returns the largest index M
!     ARR(M) =< VAL
!
!  If no such M (i.e. ARR(1) > VAL) then M=0
!
	real(SP), intent(in) :: arr(:), val
	integer(I4B) :: upper_index_sp

	integer(I4B) :: n, ml, mm, mh

	n = size(arr)

! return if array has zero length:
	if (n == 0) then
		upper_index_sp = 0
		return
	endif

! find extremal cases:
	if (arr(n) <= val) then
		upper_index_sp = n
		return
	else if (arr(1) > val) then
		upper_index_sp = 0
		return
	endif

! binary chop to find solution
	ml = 1
	mh = n
 10 mm = (ml+mh)/2
	if (mm == ml) then
		upper_index_sp = ml
		return
	endif
	if (arr(mm) > val) then
		mh = mm
	else
		ml = mm
	endif
	goto 10

	end
!---------------------------------------------------------------------------------------------
	function upper_index_i4b (arr, val)
	use type_definitions
	implicit none
!
!	Given monotonically increasing array ARR the function returns the largest index M
!     ARR(M) =< VAL
!
!  If no such M (i.e. ARR(1) > VAL) then M=0
!
	integer(I4B), intent(in) :: arr(:), val
	integer(I4B) :: upper_index_i4b

	integer(I4B) :: n, ml, mm, mh

	n = size(arr)

! return if array has zero length:
	if (n == 0) then
		upper_index_i4b = 0
		return
	endif

! find extremal cases:
	if (arr(n) <= val) then
		upper_index_i4b = n
		return
	else if (arr(1) > val) then
		upper_index_i4b = 0
		return
	endif

! binary chop to find solution
	ml = 1
	mh = n
 10 mm = (ml+mh)/2
	if (mm == ml) then
		upper_index_i4b = ml
		return
	endif
	if (arr(mm) > val) then
		mh = mm
	else
		ml = mm
	endif
	goto 10

	end
